processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java

```diff
@@ -187,11 +203,112 @@ public RowSignature resultArraySignature(final ScanQuery query)
     }
   }
 
+  /**
+   * This batches the fetched {@link ScanResultValue}s which have similar signatures and are consecutives. In best case
+   * it would return a single frame, and in the worst case, it would return as many frames as the number of results
+   * passed. Note: Batching requires all the frames to be materialized before they are propogated further and this might
+   * be improved
+   */
+  @Override
+  public Optional<Sequence<FrameSignaturePair>> resultsAsFrames(
+      final ScanQuery query,
+      final Sequence<ScanResultValue> resultSequence,
+      @Nullable Long memoryLimitBytes
+  )
+  {
+    final AtomicLong memoryLimitAccumulator = memoryLimitBytes != null ? new AtomicLong(memoryLimitBytes) : null;
+    Yielder<ScanResultValue> yielder = Yielders.each(resultSequence);
+    RowSignature prevSignature = null;
+    List<Cursor> unwrittenCursors = null;
+    List<FrameSignaturePair> frameSignaturePairs = new ArrayList<>();
```

Comment:
Can we add final here

Reply:
All of this is prematerialized. We need to make this lazy. For next() of the iterable, we always return one frame. That frame is generated lazily.